xm-commons-lep/src/main/java/com/icthh/xm/commons/lep/XmLepScriptConfigServerResourceLoader.java

```diff
@@ -1,142 +1,196 @@
-        if (isDeleted) {
-            LOGGER.info("LEP script deleted by config path: {}", updatedKey);
-            // delete
-            scriptResources.remove(updatedKey);
-        } else {
-            scriptResources.compute(updatedKey, (key, currentValue) -> {
-                if (currentValue != null) {
-                    LOGGER.info("LEP script updated by config path: {}", updatedKey);
-                    // update
-                    currentValue.update(configContent, getCurrentMilli());
-                    return currentValue;
-                } else {
-                    LOGGER.info("LEP script created by config path: {}", updatedKey);
-                    // create
-                    return new XmLepScriptResource(updatedKey, configContent, getCurrentMilli());
-                }
-            });
+    private void refreshImmediatelyIfSynchronousMode(String updatedKey) throws InterruptedException, ExecutionException {
+        if (SYNCHRONOUS.equals(lepUpdateMode)) {
+            Set<String> tenantsToUpdate = getTenantsToUpdate(List.of(updatedKey));
+            refreshEngines(tenantsToUpdate, false).get();
+            // if refresh operation invoked in thread where inited threadLepContext, threadLepContext have to be reinited
+            LepExecutorResolver currentLepExecutorResolver = lepManagementService.getCurrentLepExecutorResolver();
+            if (currentLepExecutorResolver != null && tenantContextHolder.getContext().isInitialized()) {
+                lepManagementService.endThreadContext();
+                lepManagementService.beginThreadContext();
+            }
         }
     }
 
     @Override
+    @SneakyThrows
     public void refreshFinished(Collection<String> paths) {
-        if (fullRecompileOnLepUpdate) {
-            this.cacheableEngines.forEach(CacheableLepEngine::clearCache);
-            scriptResources.forEach((key, value) -> {
-                value.modifiedNow();
-            });
+        Set<String> tenantsToUpdate = getTenantsToUpdate(paths);
+        Future<?> future = refreshEngines(tenantsToUpdate, false);
+        if (SYNCHRONOUS.equals(lepUpdateMode)) {
+            future.get();
+        }
+    }
+
+    private Future<?> refreshEngines(Set<String> tenantsToUpdate, boolean isInit) {
+        log.info("Submit task for update lep engines for tenants {}", tenantsToUpdate);
+        return refreshExecutor.submit(() -> {
+            try {
+                // if not init - it`s refresh, if it`s init we need to check that not inited yet
+                if (isInit && lepManagementService.isLepEnginesInited()) {
+                    return false;
+                }
+
+                Map<String, List<XmLepConfigFile>> configToUpdate = prepareConfigs(tenantsToUpdate);
+                lepManagementService.refreshEngines(configToUpdate);
+                return true;
+            } catch (Throwable e) {
+                log.error("Error during refresh configs", e);
```

Comment:
seems this log statement will be better and show the reason in first line:
  log.error("Error during refresh configs: {}", e.getMessage(), e);